robot/impl/local_robot.go

```diff
@@ -325,7 +340,11 @@ func (r *localRobot) Status(ctx context.Context, resourceNames []resource.Name)
 					return nil, errors.Wrapf(err, "failed to get status from %q", name)
 				}
 			}
-			resourceStatus = robot.Status{Name: name, Status: status}
+			resourceStatus = robot.Status{
+				Name:             name,
+				LastReconfigured: lastReconfiguredMap[name],
```

Comment:
I don't think it's possible for `lastReconfiguredMap` to _not_ contain the resource we're getting the status of here. If it were possible, this would return the zero value of `time.Time`, which is January 1st, 1 AD, 00:00:00.

Reply:
Actually I'm just confused by the existing `localRobot.Status` function. This return value -- using the names of the variables -- only outputs the statuses of remote resources. I don't see why non-remote resources would be included.

Or is this function playing some trick treating itself as a "remote" to simplify the coding?

Reply:
> Isn't it simpler to just call *r.manager.resources.Node(name).LastReconfigured()? 

We could! We need the mutex to access `r.manager`, so I used this map strategy to keep the locking only in the top of the function and grab all the info we needed from the resource graph while we were building up the list of resource names and the map of those names to actual resources.

> Oh I see it now. We build up the remotes statuses first. This "!ok" clause is where all local statuses get computed.

Yep! The `Status` function is a little hard to read, but this is where we add in `LastReconfigured` for remote resources. If I'm not mistaken, remote resources will have `LastReconfigured` values determined by the remotes and not by the main robot. Let me add some more tests for `LastReconfigured` values for remotes so I can both explain my understanding and test that it's correct.

Reply:
> We need the mutex to access `r.manager`, so I used this map strategy

Ah the mutex is important. FWIW, I don't think we need the mutex to access `r.manager`. The [node lookup](https://github.com/viamrobotics/rdk/blob/de7417753fcbdd420a5d4e281e2ada37f45d75d0/resource/resource_graph.go#L174-L177) gets a mutex for safety of its internal data structures. And the same goes for the `GraphNode` getters. But we do want the resource graph to stay put while we query it, get a single consistent snapshot. (I presume all modifications to the resource graph happen while holding the robot mutex.

Thinking about that some more, is there any guarantee [this "status" call succeeds](https://github.com/viamrobotics/rdk/blob/de7417753fcbdd420a5d4e281e2ada37f45d75d0/robot/impl/local_robot.go#L322)? Is it possible that the resource no longer exists because a reconfigure happened that dropped it?

I'm fine with the changes you have, but if you're interested in being more aggressive in refactoring the status function, I'd be open to seeing what comes out of that. And I'd be willing to risk writing some bugs in the process.

Reply:
> (I presume all modifications to the resource graph happen while holding the robot mutex)... Is it possible that the resource [we're querying for status] no longer exists because a reconfigure happened that dropped it?

Actually, `localRobot.mu` is _only_ used in `Status`. I believe it used to be used for `Close`, but you removed those usages. I agree that we don't need much "extra" locking on top of the existing locks on the resource manager (`configLock` guarding `updateResources` and `completeConfig`), the resource graph (`mu` guarding most access), and graph nodes (`mu` guarding most access). But, I also think that it was possible that a reconfigure could come in and drop a resource mid `Status`-calculation. I've renamed `localRobot.mu` to `localRobot.statusMu` and used it around the entire `Status` method.

I also refactored the `Status` method slightly. Just left a few more comments and renamed some variables for clarity; I believe you wrote some of the original code @cheukt. Let me know if anything looks weird I'm not particularly attached to any of my changes, was mostly trying to document my own understanding.

Still working on another test for remote resource `lastReconfigured` behavior.

Reply:
changes look good 👍